# Analysis results generated by PAPipe

1. Trimmed read data

    - QC results for all samples before trimming
        
        ```
        Path_to_out_directory/00_ReadQC/QC_Report_Before_Trimming/[population]/[sample]_1_fastqc.html
        Path_to_out_directory/00_ReadQC/QC_Report_Before_Trimming/[population]/[sample]_2_fastqc.html
        
        ```

    - MultiQC summarized QC results for populations before trimming
        
        ```
        /Path_to_out_directory/00_ReadQC/QC_Report_Before_Trimming/[population]/multiqc_report.html
        
        ```
        
    - QC results for all samples after trimming
        
        ```
        Path_to_out_directory/00_ReadQC/QC_Report_After_Trimming/[population]/[sample]_1_fastqc.html
        Path_to_out_directory/00_ReadQC/QC_Report_After_Trimming/[population]/[sample]_2_fastqc.html
        
        ```

    - MultiQC summarized QC results for populations after trimming
        
        ```
        /Path_to_out_directory/00_ReadQC/QC_Report_After_Trimming/[population]/multiqc_report.html
        
        ```

    - Trimmed read data for each sample
        
        ```
        Path_to_out_directory/00_ReadQC/TrimmedData/[sample]_1_val_1.fq.gz
        Path_to_out_directory/00_ReadQC/TrimmedData/[sample]_2_val_2.fq.gz
        
        ```
        
2. Read alignment data
    - Read mapping files for all samples
        
        ```
        Path_to_out_directory/01_readMapping/04.ReadRegrouping/[population]_[sample].addRG.marked.sort.bam
        
        ```
        
3. Variant call data
    - Variant call results generated using all population sequencing data
        
        ```
        Path_to_out_directory/02_VariantCalling/VariantCalling/All.variant.combined.GT.SNP.flt.vcf.gz
        
        ```
        
3. Post-processed data
    - Variant call gone through Hapmap format conversion/Plink filtering
        
        ```
        Path_to_out_directory/03_Postprocessing/Hapmap/All.variant.combined.GT.SNP.flt.hapmap
        Path_to_out_directory/03_Postprocessing/plink/ALL.*
        
        ```
        
4. Population analysis
   
    (1). Principal component analysis by PLINK v 1.9
   
        - Main result directory

            ```
            Path_to_out_directory/04_Population/[running datetime]/PCA/
            ```
   
        - PCA results
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/PCA/PCs.info
            ```
            
        - PCA plots of the combination of two PCs
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/PCA/all.PCA.pdf
            ```
            
    (2). Principal component analysis by PLINK v 2.0
   
        - Main result directory
            ```
            Path_to_out_directory/04_Population/[running datetime]/Plink2/
            ```
   
        - PCA results
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/Plink2/PCs.info
            ```
   
        - PCA plots of the combination of two PCs
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/Plink2/all.PCA.pdf
            ```
            
    (3). Phylogenetic tree analysis
   
        - Main result directory
            ```
            Path_to_out_directory/04_Population/[running datetime]/PhylogeneticTree/
            
            ```
   
        - .NEWICK formatted phylogenetic tree
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/PhylogeneticTree/snphylo.ml.txt
            ```
            
        - Visualized phylogenetic tree
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/PhylogeneticTree/snphylo.ml.png
            ```
            
    (4). Population tree analysis
   
        - Main result directory
            ```
            Path_to_out_directory/04_Population/[running datetime]/Treemix/
            ```
   
        - Treemix results in a single PDF file
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/Treemix/Treemix.results.pdf
            ```
            
    (5). Population structure analysis
   
        - Main result directory
            ```
            Path_to_out_directory/04_Population/[running datetime]/Structure/
            ```
   
        - STRUCTURE results per K in .PNG files
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/Structure/CLUMPAK/K=[n].MajorCluster.png
            ```
            
        - STRUCTURE results for all K in single .PDF file
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/Structure/CLUMPAK/job_100_pipeline_summary.pdf
            ```
            
    (6). Linkage disequilibrium decay analysis
   
        - Main result directory
            ```
            Path_to_out_directory/04_Population/[running datetime]/LdDecay/
            ```
   
        - Linkage disequilibrium decay results for each maximum distance parameter
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/LdDecay/[maxDist]/Plot/out.pdf
            ```
            
    (7). Selective sweep analysis
   
        - Main result directory
            ```
            Path_to_out_directory/04_Population/[running datetime]/SweepFinder2/
            ```
   
        - Selective Sweep results in point plot figures for all chromosom generated per population
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/SweepFinder2/[population]/SweepFinderOut.pdf
            ```
            
        - Selective Sweep results per population and per chromosome
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/SweepFinder2/[population]/[population].[chromosome].SF2out
            ```
            
    (8). Population admixture analysis
   
        - Main result directory
            ```
            Path_to_out_directory/04_Population/[running datetime]/AdmixtureProportion/
            ```
   
        - Admixture analysis results
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/AdmixtureProportion/admixtools_3pop/3pop.out
            Path_to_out_directory/04_Population/[running datetime]/AdmixtureProportion/admixtools_4diff/4diff.out
            Path_to_out_directory/04_Population/[running datetime]/AdmixtureProportion/admixtools_f4stat/f4stat.out
            Path_to_out_directory/04_Population/[running datetime]/AdmixtureProportion/admixtools_Dstat/Dstat.out
            ```
            
    (9). Pairwise sequentially Markovian coalescent analysis
   
        - Main result directory
            ```
            Path_to_out_directory/04_Population/[running datetime]/EffectiveSize/
            ```
   
        - Effective Size plot
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/EffectiveSize/psmc_plot.pdf
            ```
            
    (10). Multiple sequentially Markovian coalescent analysis
   
        - Main result directory
            ```
            Path_to_out_directory/04_Population/[running datetime]/MSMC/
            ```
   
        - Effective Size plot
            
            ```
            Path_to_out_directory/04_Population/[running datetime]/MSMC/MSMC.pdf
            ```
            
    (11). Fixation index analysis
   
        - Main result directory
            ```
            Path_to_out_directory/04_Population/[running datetime]/Fst/
            ```
   
        - Fixation index results visualized in manhatton plot figures
            
            ```
            #pairwise comparison
            Path_to_out_directory/04_Population/[running datetime]/Fst/[comparing pair information]/[comparing pair information].pdf
            
            #single population vs all other populations, user combination
            Path_to_out_directory/04_Population/[running datetime]/Fst/[comparing information]/Fst_result.pdf
            ```
            
        - Significant regions results of Fst analysis
            
            ```
            #pairwise comparison
            /Path_to_out_directory/04_Population/[running datetime]/Fst/[comparing pair information]/[comparing pair information].sig.region.txt

            #single population vs all other populations, user combination
            /Path_to_out_directory/04_Population/[running datetime]/Fst/[comparing information]/Fst_result.sig.region.txt
            
            ```
